spec/es6.webidl

```diff
@@ -2,8 +2,10 @@
 
 typedef FrozenArray<(SpreadElement or Expression)> Arguments;
 typedef DOMString string;
-typedef string Identifier;
-typedef string IdentifierName;
+
+typedef [IdentifierDeclaration] string IdentifierDeclaration;
+
+typedef [IdentifierReference] string IdentifierReference;
```

Comment:
If you agree that these changes make sense, I'll submit them separately upstream.

What do you think?

Reply:
about declaration/reference/string, the difference is not much clear in some case,
it would need some brush up.

about scope, to my understanding it's there because we now use "*Content" interface,
and interfaces which has "Asserted*Scope" are not same as the scope's owner in some case, right?
maybe defining it as the point which the "Asserted*Scope" actually belong to would make sense.

Reply:
In both cases, I want to explicitly expose some information to the grammar reflection, much as `[TypeIndicator]` and `[NonEmpty]`, rather than hardcoding them somewhere.

For declaration/reference, the immediate application is that tokenreaders/writers know the difference between an identifier and a string, just as they know the difference between a string enum and a literal string.

For scopes, the idea is to be able to attach an identifier to a scope. I believe that this has applications in tooling, but more immediately, it lets us do e.g. DeBruijn indices, or using per-scope heuristics.

Reply:
if it's for string vs identifier, is there any reason that declaration and reference should be different?

about scope, I think it's reasonable.

Reply:
Well, if we want to do some DeBruijn-style indices, or anything that attaches an identifier to a specific scope, we need to make the difference between declaration and reference, right? I can't think of a better way to do it.